### PR TITLE
Adds DER Regularization.

### PR DESCRIPTION
**Pull request by mockoocy**
Based on the paper: https://arxiv.org/abs/2004.07211

Since DER is a strategy based on gradient regularization, the existing Model abstraction does not suffice, i.e. using Model.fit method in order to learn using DER Strategy would not make it possible to modify loss values and update buffer during training of the model.
To circumvent this, TorchBackedModel abstract class has been added, which exposes the necessary model internals.

Alongside the strategy, reservoir sampling has been added via ReservoirBuffer.

---
Note: When running tests for these changes, I've  noticed that some tests don't pass (ones related to `MemoryUsageCallback`). It also occurs for the current `main` branch.

### CODE REVIEW

File: src/pyclad/strategies/regularization/der.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                    x_alpha, z_alpha, _ = self._buffer.sample(n=n, target_device=device)
+                    x_beta, _, _ = self._buffer.sample(n=n, target_device=device)
+                    alpha_step = module.training_step((x_alpha,), 0)
+                    beta_step = module.training_step((x_beta,), 0)
+                    loss = loss + self._alpha * F.mse_loss(alpha_step["output"], z_alpha)
+                    loss = loss + self._beta * beta_step["loss"]
+
+                loss.backward()
+                optimizer.step()
+                self._buffer.update(
```
Comment (Nyderx) [inline src/pyclad/strategies/regularization/der.py:85]: Should we update a buffer during each epoch or only during epoch 1? Could you please verify?

Comment (mockoocy) [reply src/pyclad/strategies/regularization/der.py:85]: Looking at the implementation mentioned in the paper: https://github.com/aimagelab/mammoth/blob/cd96b28cbe7284832aebd38d9ea534fb3d3ead10/models/der.py#L29
it seems like buffer is being updated each training step, without a check for current epoch.

**Review by Nyderx — APPROVED** [review]
